src/roaring64.c

```diff
@@ -1320,6 +1312,283 @@ void roaring64_bitmap_andnot_inplace(roaring64_bitmap_t *r1,
+        roaring_bitmap_free_without_containers(bitmap32);
+    }
+
+    return buf - initial_buf;
+}
+
+size_t roaring64_bitmap_portable_deserialize_size(const char *buf,
+                                                  size_t maxbytes) {
+    // https://github.com/RoaringBitmap/RoaringFormatSpec#extension-for-64-bit-implementations
+    if (buf == NULL) {
+        return 0;
+    }
+    const char *maxbuf;
+    if (UINTPTR_MAX - maxbytes < (uintptr_t)buf) {
+        maxbuf = (const char *)UINTPTR_MAX;
+    } else {
+        maxbuf = buf + maxbytes;
+    }
+    const char *initial_buf = buf;
+
+    // Read as uint64 the distinct number of "buckets", where a bucket is
+    // defined as the most significant 32 bits of an element.
+    uint64_t buckets;
+    if (buf + sizeof(buckets) > maxbuf) {
+        return 0;
+    }
+    memcpy(&buckets, buf, sizeof(buckets));
+    buf += sizeof(buckets);
+
+    // Buckets should be 32 bits with 4 bits of zero padding.
+    if (buckets > UINT32_MAX) {
+        return 0;
+    }
+
+    // Iterate through buckets ordered by increasing keys.
+    for (uint64_t bucket = 0; bucket < buckets; ++bucket) {
+        // Read as uint32 the most significant 32 bits of the bucket.
+        uint32_t high32;
+        if (buf + sizeof(high32) > maxbuf) {
+            return 0;
+        }
+        buf += sizeof(high32);
+
+        // Read the 32-bit Roaring bitmaps representing the least significant
+        // bits of a set of elements.
+        size_t bitmap32_size =
+            roaring_bitmap_portable_deserialize_size(buf, maxbuf - buf);
+        if (bitmap32_size == 0) {
+            return 0;
+        }
+        buf += bitmap32_size;
+    }
+    return buf - initial_buf;
+}
+
+roaring64_bitmap_t *roaring64_bitmap_portable_deserialize_safe(
+    const char *buf, size_t maxbytes) {
+    // https://github.com/RoaringBitmap/RoaringFormatSpec#extension-for-64-bit-implementations
+    if (buf == NULL) {
+        return NULL;
+    }
+    const char *maxbuf;
+    if (UINTPTR_MAX - maxbytes < (uintptr_t)buf) {
+        maxbuf = (const char *)UINTPTR_MAX;
+    } else {
+        maxbuf = buf + maxbytes;
+    }
+
+    // Read as uint64 the distinct number of "buckets", where a bucket is
+    // defined as the most significant 32 bits of an element.
+    uint64_t buckets;
+    if (buf + sizeof(buckets) > maxbuf) {
+        return NULL;
+    }
+    memcpy(&buckets, buf, sizeof(buckets));
+    buf += sizeof(buckets);
+
+    // Buckets should be 32 bits with 4 bits of zero padding.
+    if (buckets > UINT32_MAX) {
+        return NULL;
+    }
+
+    roaring64_bitmap_t *r = roaring64_bitmap_create();
+    // Iterate through buckets ordered by increasing keys.
+    for (uint64_t bucket = 0; bucket < buckets; ++bucket) {
+        // Read as uint32 the most significant 32 bits of the bucket.
+        uint32_t high32;
+        if (buf + sizeof(high32) > maxbuf) {
+            return NULL;
```

Comment:
We need to cleanup `r` in all these early return cases, or we'll leak memory

Reply:
Done.